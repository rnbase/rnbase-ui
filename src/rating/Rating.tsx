import React, { useCallback, useMemo, useState, useEffect } from 'react';
import {
  Animated,
  PanResponder,
  StyleProp,
  StyleSheet,
  TextStyle,
  View,
  ViewProps,
  ViewStyle,
} from 'react-native';

import { Themed, Theme, WithThemeProps, withTheme } from '../theming';
import { inRange } from '../helpers';
import RatingSymbol from './RatingSymbol';

interface RatingProps extends ViewProps {
  size?: number;
  type?: 'solid' | 'outline';
  value?: number;
  animate?: boolean;
  maxValue?: number;
  activeColor?: string;
  ratingColor?: string;
  underlayColor?: string;
  allowDecimals?: boolean;
  style?: StyleProp<ViewStyle>;
  onChange?: Function;
  onFinish?: Function;
}

const Rating: React.FC<Themed<typeof createStyleSheet, RatingProps>> = ({
  styles,
  size = 20,
  type = 'outline',
  value = 0,
  animate = true,
  maxValue = 5,
  activeColor,
  ratingColor,
  underlayColor,
  allowDecimals = false,
  style,
  onChange,
  onFinish,
}) => {
  const [rating, setRating] = useState();
  const [overlayWidth] = useState(() => new Animated.Value(0));
  const setOverlayWidth = useCallback(ratingValue => overlayWidth.setValue(ratingValue * size), [
    overlayWidth,
    size,
  ]);

  useEffect(() => setOverlayWidth(inRange(value, 0, maxValue)), [setOverlayWidth, value, maxValue]);

  const onMove = useCallback(
    ({ nativeEvent }) => {
      const eventValue = allowDecimals
        ? Math.round((nativeEvent.locationX / size) * 10) / 10
        : Math.ceil(nativeEvent.locationX / size);
      const ratingValue = inRange(eventValue, 0, maxValue);

      setRating(ratingValue);
      setOverlayWidth(ratingValue);

      if (onChange) {
        onChange(ratingValue);
      }
    },
    [allowDecimals, size, maxValue, setOverlayWidth, onChange],
  );

  const onRelease = useCallback(() => {
    if (onFinish) {
      onFinish(rating);
    }

    setRating(undefined);
  }, [onFinish, rating]);

  const panResponder = useMemo(() => {
    if (!onChange && !onFinish) {
      return false;
    }

    return PanResponder.create({
      onStartShouldSetPanResponder: () => true,
      onPanResponderTerminationRequest: () => false,
      onPanResponderGrant: onMove,
      onPanResponderMove: onMove,
      onPanResponderRelease: onRelease,
    });
  }, [onChange, onFinish, onMove, onRelease]);

  const rootStyles = [
    styles.root,
    style,
    {
      width: size * maxValue,
      height: size,
    },
  ];

  const overlayStyles = [
    styles.root,
    styles.overlay,
    {
      width: overlayWidth,
      height: size,
    },
  ];

  const textStyle = {
    width: size,
    fontSize: size,
    height: size * 1.25,
    lineHeight: size * 1.25,
  };

  const underlaySymbolStyles = [textStyle, styles.symbol, styles.symbolUnderlay];
  const overlaySymbolStyles = [textStyle, styles.symbol, styles.symbolOverlay];

  if (underlayColor) {
    underlaySymbolStyles.push({ color: underlayColor });
  }

  if (ratingColor) {
    overlaySymbolStyles.push({ color: ratingColor });
  }

  if (rating) {
    overlaySymbolStyles.push(activeColor ? { color: activeColor } : styles.symbolActive);
  }

  const getSymbols = useCallback(
    layer => Array(maxValue).fill(layer === 'overlay' || type === 'solid' ? '\u2605' : '\u2606'),
    [maxValue, type],
  );

  const renderSymbols = (layer: 'underlay' | 'overlay', symbolStyle: StyleProp<TextStyle>) =>
    getSymbols(layer).map((symbol, index) => (
      <RatingSymbol
        key={index}
        selected={animate && Math.ceil(rating) === index + 1}
        style={symbolStyle}
      >
        {symbol}
      </RatingSymbol>
    ));

  return (
    <View {...panResponder && panResponder.panHandlers} pointerEvents="box-only" style={rootStyles}>
      {renderSymbols('underlay', underlaySymbolStyles)}
      <Animated.View style={overlayStyles}>
        {renderSymbols('overlay', overlaySymbolStyles)}
      </Animated.View>
    </View>
  );
};

const createStyleSheet = ({ Colors }: Theme) =>
  StyleSheet.create({
    root: {
      flexShrink: 0,
      alignItems: 'center',
      flexDirection: 'row',
    },
    overlay: {
      top: 0,
      left: 0,
      overflow: 'hidden',
      position: 'absolute',
    },
    symbol: {
      flexShrink: 0,
      fontWeight: '200',
      textAlign: 'center',
      fontFamily: 'System',
      includeFontPadding: false,
      textAlignVertical: 'center',
      backgroundColor: 'transparent',
    },
    symbolUnderlay: {
      color: Colors.gray3,
    },
    symbolOverlay: {
      color: Colors.orange,
    },
    symbolActive: {
      color: Colors.red,
    },
  });

export type Props = RatingProps & WithThemeProps;

export default withTheme(Rating, createStyleSheet, 'Rating');
