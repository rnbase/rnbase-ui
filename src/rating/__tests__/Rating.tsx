import React from 'react';

// Note: test renderer must be required after react-native.
import TestRenderer, { act } from 'react-test-renderer';

/**
 * Under test
 */
import Rating, { Props } from '../Rating';

jest.mock('../RatingSymbol', () => 'RatingSymbol');

const createElement = (props: Props) => <Rating {...props} />;

const createRenderer = (props: Props) => TestRenderer.create(createElement(props));

it('should render normally', () => {
  const tree = createRenderer({ value: 3 });

  act(() => {});

  expect(tree).toMatchSnapshot();
});

it('should render with custom props', () => {
  const tree = createRenderer({
    value: 3,
    size: 30,
    type: 'solid',
    maxValue: 10,
  });

  act(() => {});

  expect(tree).toMatchSnapshot();
});

it('should render with custom styles', () => {
  const tree = createRenderer({
    value: 3,
    symbolStyle: { color: 'black' },
    activeSymbolStyle: { color: 'blue' },
    selectedSymbolStyle: { color: 'green' },
  });

  act(() => {});

  expect(tree).toMatchSnapshot();
});
