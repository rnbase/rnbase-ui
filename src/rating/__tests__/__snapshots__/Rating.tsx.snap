// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render normally 1`] = `
<View
  pointerEvents="box-only"
  style={
    Array [
      Object {
        "height": 20,
        "width": 100,
      },
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexShrink": 0,
      },
      undefined,
    ]
  }
>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 20,
          "height": 25,
          "lineHeight": 25,
          "width": 20,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    ☆
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 20,
          "height": 25,
          "lineHeight": 25,
          "width": 20,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    ☆
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 20,
          "height": 25,
          "lineHeight": 25,
          "width": 20,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    ☆
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 20,
          "height": 25,
          "lineHeight": 25,
          "width": 20,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    ☆
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 20,
          "height": 25,
          "lineHeight": 25,
          "width": 20,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    ☆
  </RatingSymbol>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexShrink": 0,
        "height": 20,
        "left": 0,
        "overflow": "hidden",
        "position": "absolute",
        "top": 0,
        "width": 60,
      }
    }
  >
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 20,
              "height": 25,
              "lineHeight": 25,
              "width": 20,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            undefined,
          ],
          Object {
            "color": "#ff9500",
          },
          undefined,
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 20,
              "height": 25,
              "lineHeight": 25,
              "width": 20,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            undefined,
          ],
          Object {
            "color": "#ff9500",
          },
          undefined,
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 20,
              "height": 25,
              "lineHeight": 25,
              "width": 20,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            undefined,
          ],
          Object {
            "color": "#ff9500",
          },
          undefined,
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 20,
              "height": 25,
              "lineHeight": 25,
              "width": 20,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            undefined,
          ],
          Object {
            "color": "#ff9500",
          },
          undefined,
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 20,
              "height": 25,
              "lineHeight": 25,
              "width": 20,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            undefined,
          ],
          Object {
            "color": "#ff9500",
          },
          undefined,
        ]
      }
    >
      ★
    </RatingSymbol>
  </View>
</View>
`;

exports[`should render with custom props 1`] = `
<View
  pointerEvents="box-only"
  style={
    Array [
      Object {
        "height": 30,
        "width": 300,
      },
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexShrink": 0,
      },
      undefined,
    ]
  }
>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 30,
          "height": 37.5,
          "lineHeight": 37.5,
          "width": 30,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    ★
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 30,
          "height": 37.5,
          "lineHeight": 37.5,
          "width": 30,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    ★
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 30,
          "height": 37.5,
          "lineHeight": 37.5,
          "width": 30,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    ★
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 30,
          "height": 37.5,
          "lineHeight": 37.5,
          "width": 30,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    ★
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 30,
          "height": 37.5,
          "lineHeight": 37.5,
          "width": 30,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    ★
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 30,
          "height": 37.5,
          "lineHeight": 37.5,
          "width": 30,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    ★
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 30,
          "height": 37.5,
          "lineHeight": 37.5,
          "width": 30,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    ★
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 30,
          "height": 37.5,
          "lineHeight": 37.5,
          "width": 30,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    ★
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 30,
          "height": 37.5,
          "lineHeight": 37.5,
          "width": 30,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    ★
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 30,
          "height": 37.5,
          "lineHeight": 37.5,
          "width": 30,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    ★
  </RatingSymbol>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexShrink": 0,
        "height": 30,
        "left": 0,
        "overflow": "hidden",
        "position": "absolute",
        "top": 0,
        "width": 90,
      }
    }
  >
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 30,
              "height": 37.5,
              "lineHeight": 37.5,
              "width": 30,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            undefined,
          ],
          Object {
            "color": "#ff9500",
          },
          undefined,
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 30,
              "height": 37.5,
              "lineHeight": 37.5,
              "width": 30,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            undefined,
          ],
          Object {
            "color": "#ff9500",
          },
          undefined,
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 30,
              "height": 37.5,
              "lineHeight": 37.5,
              "width": 30,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            undefined,
          ],
          Object {
            "color": "#ff9500",
          },
          undefined,
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 30,
              "height": 37.5,
              "lineHeight": 37.5,
              "width": 30,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            undefined,
          ],
          Object {
            "color": "#ff9500",
          },
          undefined,
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 30,
              "height": 37.5,
              "lineHeight": 37.5,
              "width": 30,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            undefined,
          ],
          Object {
            "color": "#ff9500",
          },
          undefined,
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 30,
              "height": 37.5,
              "lineHeight": 37.5,
              "width": 30,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            undefined,
          ],
          Object {
            "color": "#ff9500",
          },
          undefined,
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 30,
              "height": 37.5,
              "lineHeight": 37.5,
              "width": 30,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            undefined,
          ],
          Object {
            "color": "#ff9500",
          },
          undefined,
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 30,
              "height": 37.5,
              "lineHeight": 37.5,
              "width": 30,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            undefined,
          ],
          Object {
            "color": "#ff9500",
          },
          undefined,
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 30,
              "height": 37.5,
              "lineHeight": 37.5,
              "width": 30,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            undefined,
          ],
          Object {
            "color": "#ff9500",
          },
          undefined,
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 30,
              "height": 37.5,
              "lineHeight": 37.5,
              "width": 30,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            undefined,
          ],
          Object {
            "color": "#ff9500",
          },
          undefined,
        ]
      }
    >
      ★
    </RatingSymbol>
  </View>
</View>
`;

exports[`should render with custom styles 1`] = `
<View
  pointerEvents="box-only"
  style={
    Array [
      Object {
        "height": 20,
        "width": 100,
      },
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexShrink": 0,
      },
      undefined,
    ]
  }
>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 20,
          "height": 25,
          "lineHeight": 25,
          "width": 20,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        Object {
          "color": "black",
        },
      ]
    }
  >
    ☆
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 20,
          "height": 25,
          "lineHeight": 25,
          "width": 20,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        Object {
          "color": "black",
        },
      ]
    }
  >
    ☆
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 20,
          "height": 25,
          "lineHeight": 25,
          "width": 20,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        Object {
          "color": "black",
        },
      ]
    }
  >
    ☆
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 20,
          "height": 25,
          "lineHeight": 25,
          "width": 20,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        Object {
          "color": "black",
        },
      ]
    }
  >
    ☆
  </RatingSymbol>
  <RatingSymbol
    selected={false}
    style={
      Array [
        Object {
          "fontSize": 20,
          "height": 25,
          "lineHeight": 25,
          "width": 20,
        },
        Object {
          "backgroundColor": "transparent",
          "color": "#c7c7cc",
          "flexShrink": 0,
          "fontFamily": "System",
          "fontWeight": "200",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        Object {
          "color": "black",
        },
      ]
    }
  >
    ☆
  </RatingSymbol>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexShrink": 0,
        "height": 20,
        "left": 0,
        "overflow": "hidden",
        "position": "absolute",
        "top": 0,
        "width": 60,
      }
    }
  >
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 20,
              "height": 25,
              "lineHeight": 25,
              "width": 20,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            Object {
              "color": "black",
            },
          ],
          Object {
            "color": "#ff9500",
          },
          Object {
            "color": "blue",
          },
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 20,
              "height": 25,
              "lineHeight": 25,
              "width": 20,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            Object {
              "color": "black",
            },
          ],
          Object {
            "color": "#ff9500",
          },
          Object {
            "color": "blue",
          },
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 20,
              "height": 25,
              "lineHeight": 25,
              "width": 20,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            Object {
              "color": "black",
            },
          ],
          Object {
            "color": "#ff9500",
          },
          Object {
            "color": "blue",
          },
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 20,
              "height": 25,
              "lineHeight": 25,
              "width": 20,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            Object {
              "color": "black",
            },
          ],
          Object {
            "color": "#ff9500",
          },
          Object {
            "color": "blue",
          },
        ]
      }
    >
      ★
    </RatingSymbol>
    <RatingSymbol
      selected={false}
      style={
        Array [
          Array [
            Object {
              "fontSize": 20,
              "height": 25,
              "lineHeight": 25,
              "width": 20,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#c7c7cc",
              "flexShrink": 0,
              "fontFamily": "System",
              "fontWeight": "200",
              "includeFontPadding": false,
              "textAlign": "center",
              "textAlignVertical": "center",
            },
            Object {
              "color": "black",
            },
          ],
          Object {
            "color": "#ff9500",
          },
          Object {
            "color": "blue",
          },
        ]
      }
    >
      ★
    </RatingSymbol>
  </View>
</View>
`;
